### DISCUSSION TITLE
Best practice for closing LogConfig

### DISCUSSION THREAD

**Discussion by gelanchez** [bitcraze/discussions#1538] [Q&A]
Hi,

I’m working with the FPV example and am adding functionality to close the connection and restart it. However, I’m encountering some errors (e.g. OSError and BrokenPipeError) that seem to originate from the logging framework. An additional problem comes after the second or third reconnection where stop receiving log data and have to restart the Crazyflie, although the images are correctly processed. I wonder if I should close better the framework while the CF is still connected after the stop command.

Here’s the approach I’m currently using:
```python
def stop(self):
    if self.state == State.CONNECTED:
        self.logconfig.data_received_cb.remove_callback(self.pos_data)
        self.cf.log.reset()
        self.logconfig.stop() # TODO needed?
        self.logconfig.delete() # TODO needed?
        self.cf.close_link()
    if self.image_downloader_thread:
        self.image_downloader_thread.stop()
    self.update_state(State.DISCONNECTED)
```

Many thanks,
Jose

**Comment by gemenerik** [comment]
Hi Jose,

Given the specific issues you're encountering, it might be helpful to simplify the scenario to isolate the problem more effectively. I suggest creating a standalone script that focuses solely on connecting to the Crazyflie, starting the log, stopping it, and then reconnecting. This approach can make it easier to reproduce the issue and help us understand what might be going wrong without the complexity of the full application.

**Reply by gelanchez** [reply]
Hi, @gemenerik,

I’ve simplified my code to reproduce an issue I’m encountering. Using the basiclog.py example as a guideline, I've implemented a similar process to what I use in my code. However, I’m experiencing similar errors after disconnecting from the Crazyflie, and on the second attempt, I’m unable to receive any logs.

I suspect that I’m not properly handling the disconnection or stopping the logs. Do you have any advice on how to handle these processes correctly?

```python
import logging
import time
import cflib.crtp
from cflib.crazyflie import Crazyflie
from cflib.crazyflie.log import LogConfig
from cflib.utils import uri_helper

logging.basicConfig(level=logging.INFO)

class LoggingExample:
    def __init__(self):
        self.uri = uri_helper.uri_from_env(default='tcp://192.168.4.1:5000')
        self._cf = Crazyflie(ro_cache=None, rw_cache='cache')
        self.is_connected = False
        self._cf.connected.add_callback(self._connected)
        self._cf.disconnected.add_callback(self._disconnected)

    def connect(self):
        if not self.is_connected:
            self._cf.open_link(self.uri)
            if not self._cf.link:
                print(f'Could not connect to {self.uri}')
                return
            if not hasattr(self._cf.link, 'cpx'):
                print('Not connecting with WiFi')
                self._cf.close_link()

    def _connected(self, link_uri):
        self.is_connected = True
        print(f'Connected to {link_uri}')
        self.logconfig = LogConfig(name='Position', period_in_ms=500)
        self.logconfig.add_variable('stabilizer.roll')

        try:
            self._cf.log.add_config(self.logconfig)
            if self.logconfig.valid:
                self.logconfig.data_received_cb.add_callback(self.log_data)
                self.logconfig.error_cb.add_callback(self.log_error)
                self.logconfig.start()
            else:
                print('One or more of the variables in the configuration was not found in log TOC')
        except KeyError as e:
            print(f'Could not start log configuration, {e} not found in TOC')
        except AttributeError:
            print('Could not add Position log config, bad configuration')

    def stop(self):
        if self.is_connected:
            self.logconfig.data_received_cb.remove_callback(self.log_data)
            self.logconfig.error_cb.remove_callback(self.log_error)
            self.logconfig.stop() # TODO needed?
            self.logconfig.delete() # TODO needed?
            self._cf.log.reset()
            self._cf.close_link()

    def log_error(self, logconf, msg):
        print('Error when logging %s: %s' % (logconf.name, msg))

    def log_data(self, timestamp, data, logconf):
        print(f'[{timestamp}][{logconf.name}]: ', end='')
        for name, value in data.items():
            print(f'{name}: {value:3.3f} ', end='')
        print()

    def _disconnected(self, link_uri):
        print('Disconnected from %s' % link_uri)
        self.is_connected = False

if __name__ == '__main__':
    cflib.crtp.init_drivers()
    connections = 0
    le = LoggingExample()
    try:
        while connections < 3: # Try and connect 3 times
            time.sleep(10)
            if not le.is_connected:
                le.connect()
            else:
                connections += 1
                time.sleep(10)
                le.stop()    
    except Exception as e:    
        print(e) 
```

Many thanks,
Jose

**Reply by ataffanel** [reply]
The lib has not really been designed for the link to open and close repetitively. So my suggestion would be to destroy and re-create the ```_cf``` object for each connection, this way you are sure you start from a clean connection each time.

**Reply by gelanchez** [reply]
Thanks, @ataffanel. I've tried that with the same result. Creating the `_cf` with all its variables on `start` and deleting it on `_disconnected` doesn't help.

**Reply by ataffanel** [reply]
It is currious that it did not help, I would have expected it to fix the problem since we have been doing that in some of our demo that have been running for hours in a row at fairs.

Just to make sure we are talking about the same thing, I was suggesting to have the connect function look like this:
``` python
    def connect(self):
        if not self.is_connected:
            self.uri = uri_helper.uri_from_env(default='tcp://192.168.4.1:5000')
            self._cf = Crazyflie(ro_cache=None, rw_cache='cache')
            self._cf.open_link(self.uri)
            if not self._cf.link:
                print(f'Could not connect to {self.uri}')
                return
            if not hasattr(self._cf.link, 'cpx'):
                print('Not connecting with WiFi')
                self._cf.close_link()
```

This would make sure a new fresh cf object is used each time.

**Reply by gelanchez** [reply]
Hi, @ataffanel, thanks and sorry for my late reply, I'm having some issues with the flashing procedure (I've opened another discussion). Yes, my connect function is that, the only difference is the addition of the callbacks, which doesn't matter here. This is the full code:

```python
import logging
import time

import cflib.crtp
from cflib.crazyflie import Crazyflie
from cflib.crazyflie.log import LogConfig
from cflib.utils import uri_helper

logging.basicConfig(level=logging.INFO)

class LoggingExample:
    def __init__(self):
        self.uri = uri_helper.uri_from_env(default='tcp://192.168.4.1:5000')
        self.is_connected = False

    def connect(self):
        if not self.is_connected:
            self._cf = Crazyflie(ro_cache=None, rw_cache='cache')
            self._cf.connected.add_callback(self._connected)
            self._cf.disconnected.add_callback(self._disconnected)
            self._cf.open_link(self.uri)
            if not self._cf.link:
                print(f'Could not connect to {self.uri}')
                return
            if not hasattr(self._cf.link, 'cpx'):
                print('Not connecting with WiFi')
                self._cf.close_link()

    def _connected(self, link_uri):
        self.is_connected = True
        print(f'Connected to {link_uri}')
        self.logconfig = LogConfig(name='Position', period_in_ms=500)
        self.logconfig.add_variable('stabilizer.roll')

        try:
            self._cf.log.add_config(self.logconfig)
            if self.logconfig.valid:
                self.logconfig.data_received_cb.add_callback(self.log_data)
                self.logconfig.error_cb.add_callback(self.log_error)
                self.logconfig.start()
            else:
                print('One or more of the variables in the configuration was not found in log TOC')
        except KeyError as e:
            print(f'Could not start log configuration, {e} not found in TOC')
        except AttributeError:
            print('Could not add Position log config, bad configuration')

    def stop(self):
        if self.is_connected:
            self.logconfig.data_received_cb.remove_callback(self.log_data)
            self.logconfig.error_cb.remove_callback(self.log_error)
            self.logconfig.stop() # TODO needed?
            self.logconfig.delete() # TODO needed?
            self._cf.log.reset()
            self._cf.close_link()

    def log_error(self, logconf, msg):
        print('Error when logging %s: %s' % (logconf.name, msg))

    def log_data(self, timestamp, data, logconf):
        print(f'[{timestamp}][{logconf.name}]: ', end='')
        for name, value in data.items():
            print(f'{name}: {value:3.3f} ', end='')
        print()

    def _disconnected(self, link_uri):
        print('Disconnected from %s' % link_uri)
        self.is_connected = False
        del self._cf


if __name__ == '__main__':
    cflib.crtp.init_drivers()
    connections = 0
    le = LoggingExample()
    try:
        while connections < 3: # Try and connect 3 times
            time.sleep(10)
            if not le.is_connected:
                le.connect()
            else:
                connections += 1
                time.sleep(10)
                le.stop()    
    except Exception as e:    
        print(e)
```

I'm able connect to the Crazyflie the three times, but I only receive the roll information during the first one. I wonder if there is something I need to send to the CF or close on the client side, or if it has to do with the firmware I've flashed. The only way to get the information back is by restarting the Crazyflie.

**Reply by gemenerik** [reply]
I tried your attached code a couple of times, and it has worked for me when using `self._cf.fully_connected.add_callback(self._connected)`,  however inconsistent. I suspect it has more to do with connecting over wifi & cpx than it being a problem with the logconfig.

**Reply by gelanchez** [reply]
Hi, @gemenerik. I've tried with the `fully_connected` method and the result is the same. I agree this looks more a problem of the WiFi & CPX than the logconfig, since the only way of seeing the logs again is by rebooting the crazyflie. I'm using the _wifi_img_streamer_ example there. Is there anything we can change in the Gap8 code to make it work? Nothing in the client?

Thanks,
Jose

**Reply by ToveRumar** [reply]
Hi @gelanchez !

After looking into the wifi streaming part of the AI deck code, 
it seems like we are never closing the socket, which will result in the issue you are seeing. 

We are also not taking care of other issues that could occur during sending, even though this is uncommon to bump into other issues it could happen. I think this will cause complete stop in the flow of data. 

Im creating an issue here https://github.com/bitcraze/aideck-esp-firmware/issues/24.

If you need to fix this ASAP, I would recommend trying to fix this issue by calling close() on the socket after explicit disconnect or after getting error on receive or send. This allows a new socket to take its place if you re.connect again from the client side. 

I hope this can help you to move forward with this!

**Reply by gelanchez** [reply]
Hi, @ToveRumar!

Thanks for confirming that it was an issue with the socket handling. I'll subscribe to that issue to see how it's fixed. Shouldn't you open another one in the aideck-gap8-examples repository?

I'm only handling the connection in the gap8 code as it's done in the _wifi_img_streamer_, and I'm not sure if I can do something with it to help fixing it, as I'm now dealing directly with sockets:

```c
void rx_task(void *parameters)
{
    (void)parameters; // Inform the compiler parameters are not used to avoid warnings
    char *taskname = pcTaskGetName(NULL);
    cpxPrintToConsole(LOG_TO_CRTP, "Task %s created\n", taskname);

    while (true)
    {
        // TODO client alive, multiple clients?
        cpxReceivePacketBlocking(CPX_F_WIFI_CTRL, &g_rxPacket);

        WiFiCTRLPacket_t *wifiCtrl = (WiFiCTRLPacket_t*) g_rxPacket.data; // Pointers to the same memory location

        switch (wifiCtrl->cmd)
        {
            case WIFI_CTRL_STATUS_WIFI_CONNECTED: // Not used in access point (I think)
                cpxPrintToConsole(LOG_TO_CRTP, "WiFi connected (%u.%u.%u.%u)\n",
                                wifiCtrl->data[0], wifiCtrl->data[1],
                                wifiCtrl->data[2], wifiCtrl->data[3]);
                g_wifiConnected = true;                
                break;
            case WIFI_CTRL_STATUS_CLIENT_CONNECTED:
                cpxPrintToConsole(LOG_TO_CRTP, "WiFi client connection status: %u\n", wifiCtrl->data[0]);
                g_wifiClientConnected = true;
                break;
            default:
                break;
        }
    }
}
```
I might need to wait for the fix or an example of better handling.

Thanks for all,
Jose
